ext/standard/array.c

```diff
@@ -2819,15 +2837,20 @@ PHP_FUNCTION(range)
 	}
 
 	/* If the range is given as strings, generate an array of characters. */
-	if (start_type == IS_STRING || end_type == IS_STRING) {
-		if (UNEXPECTED(start_type != end_type)) {
-			if (start_type != IS_STRING) {
-				php_error_docref(NULL, E_WARNING, "Argument #1 ($start) must be a string if argument #2 ($end)"
-					" is a string, argument #2 ($end) converted to 0");
+	if (start_type >= IS_STRING || end_type >= IS_STRING) {
+		/* If one of the inputs is NOT a string */
+		if (UNEXPECTED(start_type + end_type < 2*IS_STRING)) {
```

Comment:
This is _probably_ fine for possible values (long, array, string), but can we make this explicit, like `((1 << start_type) | (1 << end_type)) != MAY_BE_STRING`?

Reply:
Not sure why I marked three lines instead of two, I was mainly referring to the if in 2842.

Reply:
Either way, this still causes an issue if both inputs are string digits as ``start_type == IS_ARRAY`` and ``end_type == IS_ARRAY`` means that ``((1 << start_type) | (1 << end_type))`` will never be equal to ``MAY_BE_STRING``.

I thought about using the ``MAY_BE_*`` constants and do some bit mask checks but found it simpler to just use a new value that is greater than ``IS_STRING``. I can however tweak it if necessary.